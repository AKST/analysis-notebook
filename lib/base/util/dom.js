/**
 * @param {Node} parent
 * @param {Node} child
 */
export function insertAtStart(parent, child) {
  if (parent.firstChild) {
    parent.insertBefore(child, parent.firstChild);
  } else {
    parent.appendChild(child);
  }
}

/**
 * @param {HTMLElement | undefined} element
 * @param {number | boolean | string | null | undefined} value
 */
export function setDisable(element, value) {
  if (element == null) return;
  switch (value) {
    case undefined:
    case null:
    case '':
    case false:
    case 0:
      element.removeAttribute('disabled');
      break;
    default:
      element.setAttribute('disabled', 'disabled');
  }
}
