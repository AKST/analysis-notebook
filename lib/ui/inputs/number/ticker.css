div {
  position: relative;
  width: 100%;
  height: var(--input-height);
}

input {
  display: block;
  box-sizing: border-box;
  min-width: 0;
  width: var(--width);
  height: var(--input-height);
  margin: 0;
  padding: var(--input-padding);
  border: var(--border-inset-grey);
  border-width: var(--border-width);
  font-family: inherit;
}

input.webkit-input::-webkit-inner-spin-button,
input.webkit-input::-webkit-outer-spin-button {
  -webkit-appearance: none;
  border: var(--border-outset-grey);
  background: var(--bg-msgrey);
  opacity: 1;
  height: var(--input-height);
  top: 0;
  bottom: 0;
  border-width: var(--border-width);
  box-sizing: border-box;
  position: absolute;
  width: var(--spin-width);
  right: 0;
}

span {
  opacity: var(--span-opacity);
  margin: 0;
  padding: 0;
  background: var(--bg-msgrey);
  box-sizing: border-box;
  display: block;
  position: absolute;
  right: 0;
  height: calc(var(--input-height) / 2);
  width: var(--spin-width);
  border: var(--border-outset-grey);
  border-width: var(--border-width);
  pointer-events: none;
  line-height: 0.7;
  color: black;
  font-size: 9px;
  font-family: inherit;
  text-align: center;
}

#a {
  bottom: calc(var(--input-height) / 2);
  border-top-color: transparent;
}

#b {
  top: calc(var(--input-height) / 2);
  border-bottom-color: transparent;
}
