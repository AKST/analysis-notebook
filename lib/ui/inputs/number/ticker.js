import { clamp } from '../../../base/util/number.js';
import { generateCSS, updateOn } from '../../../base/dom_ui/index.js';
import { isChrome } from '../../../base/util/xplatform.js';

function getStyles() {
  const padding = 4;
  const height = 22;
  const borderWidth = 2;
  const spinWidth = 16;
  const spinGap = borderWidth + 2;

  let width;
  let spanOpacity;
  let spinnerStyle;
  if (isChrome()) {
    width = '100%';
    spanOpacity = '0';
    spinnerStyle = undefined;
  } else {
    width = `calc(100% - ${spinWidth+spinGap}px)`;
    spanOpacity = '1';
    spinnerStyle = {
      '-webkit-appearance': 'none',
      border: 'var(--border-outset-grey)',
      background: 'var(--bg-msgrey)',
      opacity: 1,
      height,
      top: 0,
      bottom: 0,
      borderWidth,
      boxSizing: 'border-box',
      position: 'absolute',
      width: spinWidth,
      right: 0,
    };
  }

  return generateCSS({
    'div': {
      position: 'relative',
      width: '100%',
      height,
    },
    'input': {
      display: 'block',
      boxSizing: 'border-box',
      minWidth: 0,
      width,
      height,
      margin: 0,
      padding,
      border: 'var(--border-inset-grey)',
      borderWidth,
      fontFamily: 'inherit',
      '&::-webkit-inner-spin-button, &::-webkit-outer-spin-button': spinnerStyle,
    },
    'span': {
      opacity: spanOpacity,
      margin: 0,
      padding: 0,
      background: 'var(--bg-msgrey)',
      boxSizing: 'border-box',
      display: 'block',
      position: 'absolute',
      right: 0,
      height: (height/2),
      width: spinWidth,
      border: 'var(--border-outset-grey)',
      borderWidth,
      pointerEvents: 'none',
      lineHeight: '0.7',
      color: 'black',
      fontSize: 9,
      fontFamily: 'inherit',
      textAlign: 'center',
    },
    '#a': {
      bottom: (height/2),
      borderTopColor: 'transparent',
    },
    '#b': {
      top: (height/2),
      borderBottomColor: 'transparent',
    },
  });
}

export class NumberTickerInput extends HTMLElement {
  /** @type {HTMLInputElement | null} */
  #input = null;

  #initialValue = 0;
  /** @type {number | null} */
  #min = null;
  /** @type {number | null} */
  #max = null;
  /** @type {number | 'any'} */
  #step = 'any';

  /**
   * @param {number} value
   * @param {number | null} min
   * @param {number | null} max
   * @param {number | 'any'} [step='any']
   */
  constructor(value, min, max, step = 'any') {
    super();
    this.attachShadow({ mode: 'open' });
    this.#initialValue = value;
    this.#min = min;
    this.#max = max;
    this.#step = step;
  }

  /**
   * @returns {number}
   */
  get value () {
    const value = this.#input?.value
    return typeof value === 'string'
      ? parseFloat(value)
      : (value ?? 0)
  }

  connectedCallback() {
    this.render();
    this.shadowRoot?.addEventListener('change', e => {
      this.dispatchEvent(new CustomEvent('change', {
        ...e,
        detail: { value: this.value },
        bubbles: true,
      }));
    });
  }

  render() {
    updateOn(this.shadowRoot, [
      ['style', null, [getStyles()]],
      ['div', null, [
        ['input', {
          type: 'number',
          value: this.#initialValue,
          min: this.#min,
          max: this.#max,
          step: this.#step,
        }, []],
        ['span', { id: 'a' }, ['+']],
        ['span', { id: 'b' }, ['-']],
      ]],
    ]);

    this.#input = /** @type {HTMLInputElement | null} */ (this.shadowRoot?.querySelector('input'));

  }

  /**
   * @param {number} value
   */
  setValue(value) {
    this.#setValue(value);
    this.render();
  }

  /**
   * @param {number} value
   */
  #setValue(value) {
    if (!this.#input) return;
    this.#input.value = String(clamp(value, this.#min, this.#max));
  }
}

customElements.define('input-number-ticker', NumberTickerInput);
