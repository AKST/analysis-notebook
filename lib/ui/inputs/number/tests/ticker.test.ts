import { expect, describe, it, beforeEach, vi } from 'vitest';
import userEvent from '@testing-library/user-event';
import { getByDisplayValue } from '@testing-library/dom';
import { NumberTickerInput } from '../ticker.js';

describe('NumberTickerInput', () => {
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    document.body.innerHTML = '';
    user = userEvent.setup();
  });

  const create = (value = 0,
                  min: number | null = null,
                  max: number | null = null,
                  step: string | number = 'any') => {
    // @ts-ignore - tests
    const tickerInput = new NumberTickerInput(value, min, max, step);
    document.body.appendChild(tickerInput);
    return tickerInput;
  };

  it('renders number ticker input correctly', () => {
    const tickerInput = create(42, 0, 100, 1);
    expect(tickerInput.shadowRoot?.innerHTML).toMatchSnapshot();
  });

  it('responds to user typing in number input', async () => {
    const tickerInput = create(5);
    const changeHandler = vi.fn();
    tickerInput.addEventListener('change', changeHandler);

    // @ts-ignore - tests
    const numberInput = getByDisplayValue(tickerInput.shadowRoot!, '5');

    // Simulate user editing the input directly
    // @ts-ignore - tests
    numberInput.value = '42';
    numberInput.dispatchEvent(new Event('change', { bubbles: true }));

    expect(changeHandler).toHaveBeenCalled();
    expect(tickerInput.value).toBe(42);
  });

  it('updates input when value is programmatically set', () => {
    const tickerInput = create(10);

    tickerInput.setValue(25);

    // @ts-ignore - tests
    const numberInput = getByDisplayValue(tickerInput.shadowRoot!, '25');
    expect(numberInput).toBeTruthy();
    expect(tickerInput.value).toBe(25);
  });
});
