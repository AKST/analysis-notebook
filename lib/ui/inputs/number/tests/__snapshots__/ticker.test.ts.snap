// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NumberTickerInput > renders number ticker input correctly 1`] = `
"<style>div {
  position: relative;
  width: 100%;
  height: 22px;
}

input {
  display: block;
  box-sizing: border-box;
  min-width: 0;
  width: calc(100% - 20px);
  height: 22px;
  margin: 0;
  padding: 4px;
  border: var(--border-inset-grey);
  border-width: 2px;
  font-family: inherit;
  &::-webkit-inner-spin-button, &::-webkit-outer-spin-button {
    -webkit-appearance: none;
    border: var(--border-outset-grey);
    background: var(--bg-msgrey);
    opacity: 1px;
    height: 22px;
    top: 0;
    bottom: 0;
    border-width: 2px;
    box-sizing: border-box;
    position: absolute;
    width: 16px;
    right: 0;
  }

}

span {
  opacity: 1;
  margin: 0;
  padding: 0;
  background: var(--bg-msgrey);
  box-sizing: border-box;
  display: block;
  position: absolute;
  right: 0;
  height: 11px;
  width: 16px;
  border: var(--border-outset-grey);
  border-width: 2px;
  pointer-events: none;
  line-height: 0.7;
  color: black;
  font-size: 9px;
  font-family: inherit;
  text-align: center;
}

#a {
  bottom: 11px;
  border-top-color: transparent;
}

#b {
  top: 11px;
  border-bottom-color: transparent;
}

</style><div><input type="number" value="42" min="0" max="100" step="1"><span id="a">+</span><span id="b">-</span></div>"
`;
